buildSrc/subprojects/dependency-modules/src/main/kotlin/gradlebuild/modules/extension/ExternalModulesExtension.kt

```diff
@@ -15,12 +15,232 @@
  */
 package gradlebuild.modules.extension
 
+import gradlebuild.modules.model.License
+
 
 abstract class ExternalModulesExtension {
 
     val groovyVersion = "2.5.12"
-
     val kotlinVersion = "1.3.72"
 
     fun futureKotlin(module: String) = "org.jetbrains.kotlin:kotlin-$module:$kotlinVersion"
+
+    val aetherApi = "org.sonatype.aether:aether-api"
```

Comment:
What I don't like about this is that any change to this file will invalidate the plugin classpath of all projects, meaning invalidating caches, rebuilding, etc... Projects which should _not_ affected by the dependency changes will be recompiled. I'm not sure this is the pattern we should encourage.

Reply:
Yes that's the old discussion. The argument for this solution would be that the GA coordinates are static and won't change. So this could, in theory, be an external plugin that provides constants for **all** libraries available on Central. Of course that's not practical, so in practice you may only have the GAs you actually need here and might need to extend the list from time to time - but **not** if you update/change versions, which should be the much more common case.

At least in the case of our build it is - we don't add new things very often, but do version updates from time to time.

Also, adding the GA is a thing you don once. With versions, you might need to try different things if you run into conflicts etc. So the process of changing a version might require several "tries", where it would be very annoying if each change leads to the complete invalidation.

That is why I think this here is a good compromise. Unless there is another good solution you can think of?

Reply:
To me the problem is that if _one_ project has to add _one_ library, then _all of them_ are invalidated, even if they don't use it. I still don't think "code completion" is a good trade-off here.

Reply:
I don't hink adding the complete coordinates as plain string in many places for dependencies you use often (like Guava) is a good solution either.

Reply:
I think not having the version is already a win, as it indeed has a higher chance of changing than adding or removing GA.

However I also agree that this is not an ideal solution. You still need to use GA coordinates for exclude for example.
I also added a number of comments around dependency entries that were not promoted to this pattern.

And if we agree that adding a GA to that file does not happen often, it also means that adding new dependencies does not happen often. So the autocomplete benefit really is useful when you are changing dependencies, which is the same moment you are changing this file and thus invalidating your whole build. So while it sounds attractive, I am afraid that in practice, pros and cons cancel themselves.

So I am on the fence on whether or not we should do this part of the change tbh. Happy to discuss further.

Reply:
> So the autocomplete benefit really is useful when you are changing dependencies, which is the same moment you are changing this file and thus invalidating your whole build. So while it sounds attractive, I am afraid that in practice, pros and cons cancel themselves.

I don't agree with this. I think it is good to have _some tool_ (being it auto-completion or something else) that helps you to see which dependencies are already use.

To get back to the example of Guava: You may use it in some of your subprojects but not in all. But then you might need it somewhere else at some point. I don't think you always change the file with the dependencies when you add a dependency in one of your subprjects. (The whole point of this is to reuse the same things in mutliple subprojects.)

In particular in this build, we more often want to use something that we already have in another place than adding something completely new.

Another use case is when you refactor your project structure (e.g. split up subprojects).

Reply:
Had a conversation with Jendrik on this and I think we should move forward with this set of changes and see how it helps / hinders in reality.